cpp/core.h

```diff
@@ -147,13 +148,23 @@ class SignalSafe {
 
   // Main thread takes signals so it can run traps.
   List<int>* TakePendingSignals() {
-    // TODO: Consider ReuseEmptyList() to avoid allocation
-    List<int>* new_queue = AllocSignalList();
     List<int>* ret = pending_signals_;
-    pending_signals_ = new_queue;
+
+    // Make sure we have a distinct list to reuse.
+    DCHECK(empty_list_ != pending_signals_);
+    pending_signals_ = empty_list_;
+
     return ret;
   }
 
+  // Main thread returns the same list as an optimization to avoid allocation.
+  void ReuseEmptyList(List<int>* empty_list) {
+    DCHECK(len(empty_list) == 0);            // main thread clears
```

Comment:
To be extra clear about assumptions here, should we add `DCHECK(empty_list->capacity_ >= kMaxPendingSignals)`? I don't know if there's a world where we decide `List<T>::clear()` should shrink the underlying slab for some reason, though (it appears we leave it the same size now). The fact that we can drop signals now means we won't accidentally allocate if the list has a tiny slab, but this might still be nice to call out? Thoughts?

Reply:
Good point, done!